### ISSUE TITLE
Intellij freezes completely when adding or removing class constraint

### ISSUE THREAD

**Issue by awalland** [carymrobbins/intellij-haskforce#228]
Main.hs
```
module Main where
import Text.Printf

main = print $ someFunc (-10000000)

someFunc :: (Num a, Ord a) => a -> a
someFunc val = if val < 0 then val * 10 else val * 9876
```

if i try to remove `, Ord a` from the class constraint intellij freezes completely with java stuck at 100%.

the same thing happens if it is initially not there and i try to add it.

curiously it works if i remove the `::`, add the class constraints and then re-add the `::`.

```
IntelliJ IDEA 15.0.3
Build #IU-143.1821, built on January 19, 2016
JRE: 1.8.0_66-b17 amd64
JVM: Java HotSpot(TM) 64-Bit Server VM by Oracle Corporation
```
```
HaskForce Version: 0.3-beta.26
```

Thank you very much in advance :)

**Comment by galderz** [comment]
I have a very similar problem. When I add a class constraint to a method, the IDE hangs on 100% :((

I've taken a thread dump which can be found [here](https://www.dropbox.com/s/j9r8s5fo3m3gz4j/intellij-hang-dump.txt?dl=0).

The following looks suspicious:

```
"AWT-EventQueue-0 IDEA 15.0.3#IU-143.1821.5 IDEA, eap:false, os:Mac OS X 10.11.2, java-version:JetBrains s.r.o 1.8.0_40-release-b119" #30 prio=6 os_prio=31 tid=0x000000010148d000 nid=0x1010b runnable [0x0000700002dba000]
   java.lang.Thread.State: RUNNABLE
	at com.intellij.extapi.psi.ASTDelegatePsiElement.getFirstChild(ASTDelegatePsiElement.java:95)
	at com.intellij.psi.impl.PsiElementBase.findChildByClass(PsiElementBase.java:301)
	at com.haskforce.psi.impl.HaskellCtypeImpl.getContext(HaskellCtypeImpl.java:27)
	at com.haskforce.psi.impl.HaskellCtypeImpl.getContext(HaskellCtypeImpl.java:13)
	at com.intellij.psi.util.PsiTreeUtil.getContextOfType(PsiTreeUtil.java:513)
	at org.jetbrains.plugins.scala.lang.psi.impl.ScalaPsiManager$CacheInvalidator$.updateModificationCount(ScalaPsiManager.scala:333)
	at org.jetbrains.plugins.scala.lang.psi.impl.ScalaPsiManager$CacheInvalidator$.childAdded(ScalaPsiManager.scala:349)
	at com.intellij.psi.impl.PsiManagerImpl.a(PsiManagerImpl.java:426)
	at com.intellij.psi.impl.PsiManagerImpl.childAdded(PsiManagerImpl.java:307)
	at com.intellij.pom.wrappers.PsiEventWrapperAspect.a(PsiEventWrapperAspect.java:94)
	at com.intellij.pom.wrappers.PsiEventWrapperAspect.update(PsiEventWrapperAspect.java:50)
	at com.intellij.pom.core.impl.PomModelImpl.runTransaction(PomModelImpl.java:180)
	- locked <0x000000076ef21058> (a com.intellij.psi.PsiLock)
	at com.intellij.psi.impl.DocumentCommitProcessor$2.run(DocumentCommitProcessor.java:224)
	- locked <0x000000076ef21058> (a com.intellij.psi.PsiLock)
	at com.intellij.psi.impl.source.codeStyle.CodeStyleManagerImpl$8.compute(CodeStyleManagerImpl.java:682)
	at com.intellij.psi.impl.source.PostprocessReformattingAspect.disablePostprocessFormattingInside(PostprocessReformattingAspect.java:124)
	at com.intellij.psi.impl.source.codeStyle.CodeStyleManagerImpl$10.compute(CodeStyleManagerImpl.java:717)
	at com.intellij.formatting.FormatterImpl.runWithFormattingDisabled(FormatterImpl.java:924)
	at com.intellij.psi.impl.source.codeStyle.CodeStyleManagerImpl.performActionWithFormatterDisabled(CodeStyleManagerImpl.java:713)
	at com.intellij.psi.impl.source.codeStyle.CodeStyleManagerImpl.performActionWithFormatterDisabled(CodeStyleManagerImpl.java:679)
	at com.intellij.psi.impl.DocumentCommitProcessor.doActualPsiChange(DocumentCommitProcessor.java:209)
	at com.intellij.psi.impl.DocumentCommitProcessor$1.process(DocumentCommitProcessor.java:145)
	at com.intellij.psi.impl.DocumentCommitProcessor$1.process(DocumentCommitProcessor.java:135)
	at com.intellij.psi.impl.PsiDocumentManagerBase.finishCommitInWriteAction(PsiDocumentManagerBase.java:333)
	at com.intellij.psi.impl.PsiDocumentManagerImpl.finishCommitInWriteAction(PsiDocumentManagerImpl.java:143)
	at com.intellij.psi.impl.PsiDocumentManagerBase$3.run(PsiDocumentManagerBase.java:298)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:1010)
	at com.intellij.psi.impl.PsiDocumentManagerBase.finishCommit(PsiDocumentManagerBase.java:295)
	at com.intellij.psi.impl.DocumentCommitThread$5.run(DocumentCommitThread.java:528)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.a(LaterInvocator.java:337)
	- locked <0x000000076eb8f590> (a java.lang.Object)
	at com.intellij.openapi.application.impl.LaterInvocator$FlushQueue.run(LaterInvocator.java:321)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:756)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:76)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:726)
	at com.intellij.ide.IdeEventQueue.c(IdeEventQueue.java:866)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:654)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:381)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)

   Locked ownable synchronizers:
	- <0x000000076a992440> (a java.util.concurrent.locks.ReentrantReadWriteLock$NonfairSync)
```

I have no idea what the Scala Language plugin is doing in that thread dump since this is a Haskell/Javascript project :\

IntelliJ version: 15.0.3
Haskforce version: 0.3-beta.26
Scala plugin (in case relevant): 2.2.0

**Comment by carymrobbins** [comment]
Ok, this is caused by a bug in Grammar-Kit, which I've filed - https://github.com/JetBrains/Grammar-Kit/issues/92

See the issue linked above for the technical details of this bug.  I think I can work around this by instructing Grammar-Kit to _not_ implement a `getContext` method.

**Comment by carymrobbins** [comment]
Ok, I should have a working fix for this.  I have been having some trouble getting travis to build successfully (mysterious since the ant build and tests work fine locally) but I'll try to get this merged into master and released by the end of the day.

Kudos to @awalland for providing a perfect test case to reproduce, and kudos to @galderz for pointing out that the Scala plugin specifically causes this bug.

**Comment by carymrobbins** [comment]
This has been resolved in the latest release, **v0.3-beta.27**.

**Comment by awalland** [comment]
great! it works! thank you very much for fixing this so quickly :)
